File: echoip.py
#!/usr/bin/env python

# ECHOIP - listen on TCP port given as argument and answer client's IP to client
# requests. It listens on all IPv6 interfaces and IPv4 requests and treaten in
# IPv6 dual stack. Usage:
#   echoip.py PORT

from socket import *
from sys import argv

def listen(address, port):
    s = socket(AF_INET6, SOCK_STREAM)
    s.setsockopt(IPPROTO_IPV6, IPV6_V6ONLY, 0)
    s.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
    s.setsockopt(SOL_SOCKET, SO_REUSEPORT, 1)
    s.bind((address, port))
    s.listen(8)

    while True:
        connect, client = s.accept()
        ip = '%s\n' % client[0].replace("::ffff:", "")
        connect.send(ip.encode())
        connect.shutdown(2)

    s.close()

if __name__ == "__main__":
    listen('', int(argv[1]))

File: echoip.pl
#!/usr/bin/env perl

# ECHOIP - listen on TCP port given as argument and answer client's IP to client
# requests. It listens on all IPv6 interfaces and IPv4 requests are treaten in
# IPv6 dual stack. Usage:
#   echoip.pl PORT

use warnings;
use strict;
use v5.024;
use IO::Socket::IP;

my $socket = IO::Socket::IP -> new(
    Listen => 8,
    V6Only => 0,
    ReuseAddr => 1,
    ReusePort => 1,
    LocalHost => '::',
    LocalPort => $ARGV[0],
    Proto => 'tcp'
) or die "IO::Socket::IP -> new(): $!\n";

while (1) {
    my $client = $socket -> accept();
    my $ipaddr = $client -> peerhost();
    $ipaddr =~ s/^::ffff://;
    $client -> send("$ipaddr\n");
    $client -> shutdown(2);
}

$socket -> close();

# vim: et sts=4 sw=4 ts=4
